packages/mdc-list/foundation.ts

```diff
@@ -295,29 +301,49 @@ export class MDCListFoundation extends MDCFoundation<MDCListAdapter> {
   }
 
   private setSingleSelectionAtIndex_(index: number) {
+    if (this.selectedIndex_ === index) return;
+
     let selectedClassName = cssClasses.LIST_ITEM_SELECTED_CLASS;
     if (this.useActivatedClass_) {
       selectedClassName = cssClasses.LIST_ITEM_ACTIVATED_CLASS;
     }
 
-    if (this.selectedIndex_ >= 0 && this.selectedIndex_ !== index) {
+    if (this.selectedIndex_ !== numbers.UNSET_INDEX) {
       this.adapter_.removeClassForElementIndex(this.selectedIndex_ as number, selectedClassName);
-      this.adapter_.setAttributeForElementIndex(this.selectedIndex_ as number, strings.ARIA_SELECTED, 'false');
     }
-
     this.adapter_.addClassForElementIndex(index, selectedClassName);
-    this.adapter_.setAttributeForElementIndex(index, strings.ARIA_SELECTED, 'true');
+    this.setAriaForSingleSelectionAtIndex_(index);
 
     this.selectedIndex_ = index;
   }
 
+  /**
+   * Sets aria attribute for single selection at given index.
+   */
+  private setAriaForSingleSelectionAtIndex_(index: number) {
+    if (this.selectedIndex_ === numbers.UNSET_INDEX) {
+      this.ariaCurrentAttrValue_ =
+            this.adapter_.getAttributeForElementIndex(index as number, strings.ARIA_CURRENT);
```

Comment:
I'm confused how this method works, since this is reading the attribute value for the element that we haven't set the attribute on yet. Wouldn't this always yield `'false'` or `null` (given that this method is never called when `selectedIndex_ === index`)? Shouldn't this be doing exactly the opposite of what it's doing - checking the attribute's value at `selectedIndex_` if `selectedIndex_` is set?

I don't see any tests verifying that `aria-current` is ever actually set to a value other than `null`, so I highly suspect a bug here.

Meanwhile, I take it this method will always fall back to using `aria-selected` if the markup had nothing selected initially? Is it not possible to detect based on the list's role instead? I don't remember whether we _required_ that an item be initially selected before.

Also, why does `index` need `as number`? It's already declared as `number`...

Reply:
To clarify your first question, `setSelectedIndex()` method is called on initialization of the component to set pre-selected items where `this.selectedIndex_` is still "unset". That's when we detect the presence of one of the aria attribute. See [component.ts:L126](https://github.com/material-components/material-components-web/blob/7357170958916857fb7f47b077496403795ad861/packages/mdc-list/component.ts#L126).

I've considered adding a new foundation method to set pre-selected index (For example, `setInitialSelectedIndex` or `setDefaultSelectedIndex`) which just sets the state - `this.selectedIndex_`, I dropped this idea since having two foundation setter methods might create ambiguity for wrapper libraries.

WDYT?

Reply:
So I thought I must've been missing something, and after talking with Abhinay, I realized I was.

Firstly, `this.ariaCurrentAttrValue_` is an instance variable, and we expect to set it exactly once here, then reuse it for all subsequent updates.

Secondly, the first call to this method (where this condition will be true) happens during the component's `initializeListType` method, at which point it calls this method with the index that _already has_ the attribute set. This is in direct contrast to subsequent calls which pass an index to select, rather than one that's already selected.

Reply:
Added a code comment to make it more clearer. Also, added new unit tests.

For records, here are the alternatives I've considered to get around detecting something from DOM during initialization to keep the state in sync with foundation:

- Move pre-selected logic from component to adapters which will be eventually called from foundation. This might need two new additional adapters - `getInitialSelectedIndex` and `setUseActivatedClass` call can be moved inside provided we've `hasClassAtIndex` that checks existence of `LIST_ITEM_ACTIVATED_CLASS` class name.
- New foundation method to set pre-selected index (For example, setInitialSelectedIndex or setDefaultSelectedIndex) which just sets the state `this.selectedIndex_`.